vllm/transformers_utils/detokenizer.py

```diff
@@ -0,0 +1,178 @@
+from vllm.transformers_utils.tokenizer_group.base_tokenizer_group import (
+    BaseTokenizerGroup)
+
+
+class Detokenizer:
+    """Provides methods to decode the output of a model into text."""
+
+    def __init__(self, tokenizer_group: BaseTokenizerGroup):
+        self.tokenizer_group = tokenizer_group
+
+    def get_tokenizer_for_seq(self,
+                              sequence: Sequence) -> "PreTrainedTokenizer":
+        """Returns the HF tokenizer to use for a given sequence."""
+        return self.tokenizer_group.get_lora_tokenizer(sequence.lora_request)
+
+    def decode_logprobs(
+            self,
+            seq: Sequence,
+            prms: SamplingParams,
+            logprobs: Dict[int, Logprob],
+            all_input_ids: List[int],
+            generated_token_id_and_text: Optional[Tuple[int,
+                                                        str]] = None) -> None:
+        """Decodes the logprobs for a sequence. In-place operation.
+
+        Args:
+            seq: The sequence to decode.
+            prms: The sampling parameters used to generate the sequence.
+            logprobs: The logprobs to decode.
+            all_input_ids: The token IDs of the sequence, including the
+                token generated this iteration.
+            generated_token_id_and_text: The token ID and text of the token
+                generated this iteration. If provided, will be used to
+                avoid decoding the token again.
+        """
+        if not logprobs:
+            return
+        tokenizer = self.get_tokenizer_for_seq(seq)
+        previous_tokens = all_input_ids[:-1]
+        for token_id, sample_logprob in logprobs.items():
+            # If the token was generated this iteration, use the provided text.
+            if (generated_token_id_and_text is not None
+                    and token_id == generated_token_id_and_text[0]):
+                sample_logprob.decoded_token = generated_token_id_and_text[1]
+                continue
+
+            if (sample_logprob.decoded_token is None and token_id != -1):
+                all_input_ids_with_logprob = previous_tokens + [token_id]
+                (_, new_text, _, _) = detokenize_incrementally(
+                    tokenizer=tokenizer,
+                    all_input_ids=all_input_ids_with_logprob,
+                    prev_tokens=seq.tokens,
+                    prefix_offset=seq.prefix_offset,
+                    read_offset=seq.read_offset,
+                    skip_special_tokens=prms.skip_special_tokens,
+                    spaces_between_special_tokens=prms.
+                    spaces_between_special_tokens,
+                )
+                sample_logprob.decoded_token = new_text
+
+    def decode_prompt_logprobs(
+            self, seq_group: SequenceGroup,
+            prompt_logprobs: List[Dict[int,
+                                       Logprob]]) -> List[Dict[int, Logprob]]:
```

Comment:
kind of odd that there's two in place operation surrounding this in place operation but also returning something, maybe just don't return here?

Reply:
also can we name everything `_inplace`

Reply:
yeah that's an artefact of previous design, will make it inplace as well and rename them